server/odc-service/src/main/java/com/oceanbase/odc/service/iam/ProjectPermissionValidator.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2023 OceanBase.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.oceanbase.odc.service.iam;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.oceanbase.odc.core.authority.model.DefaultSecurityResource;
+import com.oceanbase.odc.core.authority.permission.Permission;
+import com.oceanbase.odc.core.authority.permission.ResourceRoleBasedPermission;
+import com.oceanbase.odc.core.authority.util.SkipAuthorize;
+import com.oceanbase.odc.core.shared.constant.ResourceRoleName;
+import com.oceanbase.odc.core.shared.exception.AccessDeniedException;
+import com.oceanbase.odc.service.iam.auth.AuthenticationFacade;
+import com.oceanbase.odc.service.iam.auth.AuthorizationFacade;
+
+import lombok.NonNull;
+
+/**
+ * @author gaoda.xy
+ * @date 2024/1/23 16:38
+ */
+@Component
+public class ProjectPermissionValidator {
+
+    @Autowired
+    private AuthenticationFacade authenticationFacade;
+
+    @Autowired
+    private AuthorizationFacade authorizationFacade;
+
+    @SkipAuthorize("internal usage only")
+    public void checkProjectRole(@NonNull Long projectId, @NonNull List<ResourceRoleName> roleNames) {
+        if (!hasProjectRole(projectId, roleNames)) {
+            throw new AccessDeniedException();
+        }
+    }
+
+    @SkipAuthorize("internal usage only")
+    public void checkProjectRole(@NonNull Collection<Long> projectIds, @NonNull List<ResourceRoleName> roleNames) {
+        if (!hasProjectRole(projectIds, roleNames)) {
+            throw new AccessDeniedException();
+        }
+    }
+
+    @SkipAuthorize("internal usage only")
+    public boolean hasProjectRole(@NonNull Long projectId, @NonNull List<ResourceRoleName> roleNames) {
+        return hasProjectRole(Collections.singleton(projectId), roleNames);
+    }
+
+    @SkipAuthorize("internal usage only")
+    public boolean hasProjectRole(@NonNull Collection<Long> projectIds, @NonNull List<ResourceRoleName> roleNames) {
+        if (projectIds.isEmpty()) {
+            return false;
+        }
+        if (roleNames.isEmpty()) {
+            return true;
```

Comment:
why?

Reply:
prefer returning false
a man without any roles can pass the permission checking, it is not reasonable~

Reply:
I check all invoker, you are right, I fix it.